aesara/tensor/nlinalg.py

```diff
@@ -823,6 +824,87 @@ def tensorsolve(a, b, axes=None):
     return TensorSolve(axes)(a, b)
 
 
+class MatMul(Op):
+    __props__ = ("dtype",)
+
+    def __init__(self, dtype=None):
+        self.dtype = dtype
+
+    def make_node(self, a, b):
+        a = at.as_tensor_variable(a)
+        b = at.as_tensor_variable(b)
+
+        if 0 in (a.ndim, b.ndim):
+            raise ValueError("inputs to `matmul` cannot be scalar.")
+        elif a.ndim >= 2 and b.ndim >= 2:
+            out = at.TensorType(
+                dtype=self.dtype,
+                broadcastable=(False,) * max(a.ndim, b.ndim),
+            )()
+        else:
+            out_ndim = b.ndim - 1 if a.ndim == 1 else a.ndim - 1
+            out = at.TensorType(
+                dtype=self.dtype,
+                broadcastable=(False,) * out_ndim,
+            )()
+        return Apply(self, [a, b], [out])
+
+    def perform(self, node, inputs, outputs):
+        x1, x2 = inputs
+        outputs[0][0] = np.matmul(x1, x2, dtype=self.dtype)
+
+    def infer_shape(self, fgraph, node, shapes):
```

Comment:
I think it would be better to be "dumb" (and explicit) about the `if... elif... else` logic here.

Reply:
you're right, i made the conditionals more explicit in latest commit.